README.md
# Frontity Services

A series of serverless microservices that power up the WP REST API. They are ready to deploy with [Now](https://zeit.co/now) serverless infrastructure.

## List of services

### Add media ids

The service `/add-media-ids` searches the `content.rendered` field for images, then tries to find them in the REST API and finally attaches the results to the final response in both `content_media` array (ids) and `_embedded.wp:contentmedia` (media object).

### Add title

The service `/add-title` looks in the `title` field and converts it to `{ title: rendered: "...", text: "..." }` if it doesn't have that shape.

## Changelog

This project adheres to [Semantic Versioning](https://semver.org/) and [Angular Conventional Changelog](https://github.com/angular/angular.js/blob/master/DEVELOPERS.md#-git-commit-guidelines).
Every release is documented on the [Github Releases](https://github.com/frontity/frontity/releases) page.

## License

This project is licensed under the terms of the [Apache 2.0](https://github.com/frontity/frontity/blob/master/LICENSE) license.

## Contribute

Please take a look at our [Contribution Guide](https://github.com/frontity/contribute).


CHANGELOG.md
## [1.2.1](https://github.com/frontity/services.git/compare/v1.2.0...v1.2.1) (2018-11-29)


### Bug Fixes

* **x-wp-headers:** send WP headers from add-media-ids service ([91c9b40](https://github.com/frontity/services.git/commit/91c9b40))
* **x-wp-headers:** send WP headers from add-title service ([0fd5c27](https://github.com/frontity/services.git/commit/0fd5c27))

# [1.2.0](https://github.com/frontity/services.git/compare/v1.1.1...v1.2.0) (2018-11-22)


### Bug Fixes

* **add-title:** improves error handling ([88bbddd](https://github.com/frontity/services.git/commit/88bbddd))


### Features

* **add-title:** creates add-title service [WIP] ([55aef47](https://github.com/frontity/services.git/commit/55aef47))

## [1.1.1](https://github.com/frontity/cloud-for-demos.git/compare/v1.1.0...v1.1.1) (2018-11-12)


### Bug Fixes

* **media-request:** adds a try catch to every request ([d55b117](https://github.com/frontity/cloud-for-demos.git/commit/d55b117))

# [1.1.0](https://github.com/frontity/cloud-for-demos.git/compare/v1.0.0...v1.1.0) (2018-11-12)


### Bug Fixes

* **all:** add data-attachment-id to content [WIP] ([3d7f1ef](https://github.com/frontity/cloud-for-demos.git/commit/3d7f1ef))


### Features

* **all:** add data-attachment-id to images [WIP] ([1c22418](https://github.com/frontity/cloud-for-demos.git/commit/1c22418))
* **all:** add data-attachment-id to images [WIP] ([d92637e](https://github.com/frontity/cloud-for-demos.git/commit/d92637e))
* **all:** add data-attachment-id to images in content ([a16d75c](https://github.com/frontity/cloud-for-demos.git/commit/a16d75c))

# 1.0.0 (2018-11-05)


### Bug Fixes

* **helpers:** fix id getter from class when id is not found ([9f1e6a8](https://github.com/frontity/cloud-for-demos.git/commit/9f1e6a8))
* **helpers:** ids returned by helper are always numbers ([2c4a639](https://github.com/frontity/cloud-for-demos.git/commit/2c4a639))
* **server:** fixes definition of wp:contentmedia ([73b9734](https://github.com/frontity/cloud-for-demos.git/commit/73b9734))


### Features

* **all:** populate api calls with content_media [WIP] ([6448747](https://github.com/frontity/cloud-for-demos.git/commit/6448747))


src/__tests__/index.tests.js
/* eslint-disable no-underscore-dangle */
const nock = require('nock');
const micro = require('micro');
const got = require('got');
const listen = require('test-listen');
const server = require('../');
const listOfPosts = require('./data/list_of_posts.json');
const post10452 = require('./data/post_10452.json');
const post10452noClass = require('./data/post_10452_no_class.json');
const media11857 = require('./data/media_11857.json');
const media11848 = require('./data/media_11848.json');
const media11822 = require('./data/media_11822.json');

let service;
let url;

beforeEach(async () => {
  service = await micro(server);
  url = await listen(service);
});

afterEach(() => {
  service.close();
});

describe('Server', () => {
  test('Should throw an error if no url is passed', async () => {
    await expect(got(`${url}/`)).rejects.toThrow();
  });
  test('Should throw an error if an invalid url is passed', async () => {
    await expect(got(`${url}/an-invalid-url`)).rejects.toThrow();
  });
  test('Should return 500 status if it fails', async () => {
    try {
      await got(`${url}/not-valid-url`);
    } catch (error) {
      expect(error.statusCode).toBe(500);
    }
  });
  test('Should return 404 status if API returns 404', async () => {
    nock('https://frontity.com')
      .get('/')
      .reply(404);

    try {
      await got(`${url}/https://frontity.com`);
    } catch (error) {
      expect(error.statusCode).toBe(404);
      expect(error.statusMessage).toBe('Not Found');
    }
  });
  test('Should return a post populated with content media', async () => {
    nock('https://frontity.com')
      .get('/?rest_route=/wp/v2/posts/10452&_embed=true')
      .reply(200, post10452)
      .get('/?rest_route=/wp/v2/media/11857')
      .reply(200, media11857);

    const body = JSON.parse(
      (await got(
        `${url}/https://frontity.com/?rest_route=/wp/v2/posts/10452&_embed=true`,
      )).body,
    );

    expect(body.content_media).toEqual([11857]);
    expect(body.content.rendered).toMatch("data-attachment-id='11857'");
    expect(body._embedded['wp:contentmedia'][0][0]).toEqual(media11857);
    expect(body).toMatchSnapshot();
  });
  test("Should return a post populated with content media when images don't have class attribute", async () => {
    nock('https://frontity.com')
      .get('/?rest_route=/wp/v2/posts/10452&_embed=true')
      .reply(200, post10452noClass)
      .get('/?rest_route=/wp/v2/media&slug=stop_to_get_started-')
      .reply(200, [media11857]);

    const body = JSON.parse(
      (await got(
        `${url}/https://frontity.com/?rest_route=/wp/v2/posts/10452&_embed=true`,
      )).body,
    );

    expect(body.content_media).toEqual([11857]);
    expect(body.content.rendered).toMatch("data-attachment-id='11857'");
    expect(body._embedded['wp:contentmedia'][0][0]).toEqual(media11857);
    expect(body).toMatchSnapshot();
  });
  test('Should return a list of posts populated with content media', async () => {
    nock('https://frontity.com')
      .get('/?rest_route=/wp/v2/posts&_embed=true')
      .reply(200, listOfPosts)
      .get('/?rest_route=/wp/v2/media/11857')
      .reply(200, media11857)
      .get('/?rest_route=/wp/v2/media/11848')
      .reply(200, media11848)
      .get('/?rest_route=/wp/v2/media/11822')
      .reply(200, media11822);

    const body = JSON.parse(
      (await got(
        `${url}/https://frontity.com/?rest_route=/wp/v2/posts&_embed=true`,
      )).body,
    );

    expect(body[0].content_media).toEqual([11857]);
    expect(body[1].content_media).toEqual([11848]);
    expect(body[2].content_media).toEqual([11822]);
    expect(body[0].content.rendered).toMatch("data-attachment-id='11857'");
    expect(body[1].content.rendered).toMatch("data-attachment-id='11848'");
    expect(body[2].content.rendered).toMatch("data-attachment-id='11822'");
    expect(body[0]._embedded['wp:contentmedia'][0][0]).toEqual(media11857);
    expect(body[1]._embedded['wp:contentmedia'][0][0]).toEqual(media11848);
    expect(body[2]._embedded['wp:contentmedia'][0][0]).toEqual(media11822);
    expect(body).toMatchSnapshot();
  });
  test('Should return the same data from the original request', async () => {
    nock('https://frontity.com')
      .get('/?rest_route=/wp/v2/media/11857&_embed=true')
      .reply(200, media11857);

    const body = JSON.parse(
      (await got(
        `${url}/https://frontity.com/?rest_route=/wp/v2/media/11857&_embed=true`,
      )).body,
    );

    expect(body).toEqual(media11857);
    expect(body).toMatchSnapshot();
  });
});


src/__tests__/helpers.tests.js
const { getIdFromClass, getSlugFromSrc } = require('../helpers');

const imagesWithClass = require('./data/imagesWithClass.json');
const imagesWithClassButNoId = require('./data/imagesWithClassButNoId.json');
const imagesWithDataSrc = require('./data/imagesWithDataSrc.json');
const imagesWithSrc = require('./data/imagesWithSrc.json');
const imagesWithDataOriginal = require('./data/imagesWithDataOriginal.json');
const imagesWithSrcSizes = require('./data/imagesWithSrcSizes.json');

describe('Helper getIdFromClass', () => {
  test('Should return `null` if no id are extracted from class', () => {
    imagesWithDataSrc.forEach(image => {
      const id = getIdFromClass(image);
      expect(id).toBe(null);
    });

    imagesWithClassButNoId.forEach(image => {
      const id = getIdFromClass(image);
      expect(id).toBe(null);
    });
  });
  test('Should return an id', () => {
    imagesWithClass.forEach(image => {
      const id = getIdFromClass(image);
      expect(typeof id).toBe('number');
    });
  });
});

describe('Helper getSlugFromSrc', () => {
  test('Should return `null` if no slug are extracted', () => {
    imagesWithClass.forEach(image => {
      const slug = getSlugFromSrc(image);
      expect(slug).toBe(null);
    });
  });
  test('Should return a slug extracted from data-src', () => {
    imagesWithDataSrc.forEach(image => {
      const slug = getSlugFromSrc(image);
      expect(typeof slug).toBe('string');
    });
  });
  test('Should return a slug extracted from src', () => {
    imagesWithSrc.forEach(image => {
      const slug = getSlugFromSrc(image);
      expect(typeof slug).toBe('string');
    });
  });
  test('Should return a slug extracted from data-original', () => {
    imagesWithDataOriginal.forEach(image => {
      const slug = getSlugFromSrc(image);
      expect(typeof slug).toBe('string');
    });
  });
  test('Should return the right slug if the filename has a size', () => {
    const slugs = [
      'image_11857',
      'image_11858x',
      'image_11861-',
      'image_11859_',
      'image_11860---',
    ];

    imagesWithSrcSizes.forEach((image, index) => {
      const slug = getSlugFromSrc(image);
      expect(typeof slug).toBe('string');
      expect(slug).toBe(slugs[index]);
    });
  });
});


services/add-title/src/__tests__/index.tests.js
const nock = require('nock');
const micro = require('micro');
const axios = require('axios');
const listen = require('test-listen');
const server = require('../');
const categoryWithoutTitle = require('./data/categoryWithoutTitle.json');
const postWithTitleString = require('./data/postWithTitleString.json');
const postWithTitleRendered = require('./data/postWithTitleRendered.json');
const postWithTitleText = require('./data/postWithTitleText.json');
const listOfPostsWithTitleString = require('./data/listOfPostsWithTitleString.json');

axios.defaults.adapter = require('axios/lib/adapters/http');

let service;
let url;

beforeEach(async () => {
  service = await micro(server);
  url = await listen(service);
});

afterEach(() => {
  service.close();
});

describe('Server', () => {
  test('Should throw an error if no url is passed', async () => {
    await expect(axios(`${url}/`)).rejects.toThrow();
  });
  test('Should throw an error if an invalid url is passed', async () => {
    await expect(axios(`${url}/an-invalid-url`)).rejects.toThrow();
  });
  test('Receives a category without title and should return the original data', async () => {
    nock('https://frontity.com')
      .get('/?rest_route=/wp/v2/posts/1111125&_embed=true')
      .reply(200, categoryWithoutTitle);

    const { data } = await axios(
      `${url}/https://frontity.com/?rest_route=/wp/v2/posts/1111125&_embed=true`,
    );

    expect(data.title).toBeUndefined();
    expect(data).toMatchSnapshot();
  });
  test('Receives a post with `title` as a string and should return a post with `title.rendered` and `title.text`', async () => {
    nock('https://frontity.io')
      .get('/?rest_route=/wp/v2/posts/1111125&_embed=true')
      .reply(200, postWithTitleString);

    const title =
      'Revelados detalles y posible fecha de presentaci\u00f3n del smartphone plegable de Huawei';
    const { data } = await axios(
      `${url}/https://frontity.io/?rest_route=/wp/v2/posts/1111125&_embed=true`,
    );

    expect(data.title.rendered).toBe(title);
    expect(data.title.text).toBe(title);
    expect(data).toMatchSnapshot();
  });
  test('Receives a post with `title.rendered` and should return a post with `title.rendered` and `title.text`', async () => {
    nock('https://frontity.io')
      .get('/?rest_route=/wp/v2/posts/1111125&_embed=true')
      .reply(200, postWithTitleRendered);

    const { data } = await axios(
      `${url}/https://frontity.io/?rest_route=/wp/v2/posts/1111125&_embed=true`,
    );

    expect(data.title.rendered).toBe(
      '<h1>Revelados detalles y posible fecha de presentaci\u00f3n del smartphone plegable de Huawei</h1>',
    );
    expect(data.title.text).toBe(
      'Revelados detalles y posible fecha de presentaci\u00f3n del smartphone plegable de Huawei',
    );
    expect(data).toMatchSnapshot();
  });
  test('Receives a post with `title.text` and should return a post with `title.rendered` and `title.text`', async () => {
    nock('https://frontity.io')
      .get('/?rest_route=/wp/v2/posts/1111125&_embed=true')
      .reply(200, postWithTitleText);

    const title =
      'Revelados detalles y posible fecha de presentaci\u00f3n del smartphone plegable de Huawei';
    const { data } = await axios(
      `${url}/https://frontity.io/?rest_route=/wp/v2/posts/1111125&_embed=true`,
    );

    expect(data.title.rendered).toBe(title);
    expect(data.title.text).toBe(title);
    expect(data).toMatchSnapshot();
  });
  test('Receives a list of posts with `title` as a string and should return a list of posts with `title.rendered` and `title.text`', async () => {
    nock('https://frontity.io')
      .get('/?rest_route=/wp/v2/posts/&_embed=true&perPage=3')
      .reply(200, listOfPostsWithTitleString);

    const titles = [
      'Del deseo a la realidad: la edici\u00f3n gen\u00e9tica (a\u00fan) no est\u00e1 preparada para tratar a pacientes',
      'Ya puedes ampliar tu sistema de audio en casa con tus Chromecast',
      'Revelados detalles y posible fecha de presentaci\u00f3n del smartphone plegable de Huawei',
    ];
    const { data } = await axios(
      `${url}/https://frontity.io/?rest_route=/wp/v2/posts/&_embed=true&perPage=3`,
    );

    expect(data[0].title.rendered).toBe(titles[0]);
    expect(data[0].title.text).toBe(titles[0]);
    expect(data[1].title.rendered).toBe(titles[1]);
    expect(data[1].title.text).toBe(titles[1]);
    expect(data[2].title.rendered).toBe(titles[2]);
    expect(data[2].title.text).toBe(titles[2]);
    expect(data).toMatchSnapshot();
  });
  test('Should preserve any x-wp header', async () => {
    nock('https://frontity.com')
      .get('/?rest_route=/wp/v2/posts&_embed=true')
      .reply(200, { result: 'ok' }, { 'x-wp-custom': 'value' });
    const { headers } = await axios(
      `${url}/https://frontity.com/?rest_route=/wp/v2/posts&_embed=true`,
    );
    expect(headers['x-wp-custom']).toEqual('value');
  });
});


src/helpers.js
// Get the media ids from class attribute.
const getIdFromClass = ({ attributes }) => {
  const className = attributes.find(attribute => attribute.key === 'class');
  const idRegexp = /wp-image-(\d+)/;
  const matches = className ? className.value.match(idRegexp) : null;
  const id = matches ? matches[1] : null;

  return id && parseInt(id, 10);
};

// Get the media slugs from data-src, src or data-original attributes.
const getSlugFromSrc = ({ attributes }) => {
  const attrs = attributes.reduce(
    (result, current) => ({ ...result, [current.key]: current.value }),
    {},
  );
  const src = attrs['data-src'] || attrs.src || attrs['data-original'] || null;

  if (!src) return null;

  const filenameRegexp = /([^/\\&?]+)\.\w{2,4}(?=([?&].*$|$))/;
  const matches = src ? src.match(filenameRegexp) : null;
  const slug = matches
    ? src.match(filenameRegexp)[1].replace(/-\d+x\d+$/, '')
    : null;

  return slug;
};

module.exports = {
  getIdFromClass,
  getSlugFromSrc,
};


services/add-title/src/index.js
const { parse } = require('url');
const { send, createError } = require('micro');
const cors = require('micro-cors')();
const axios = require('axios');

module.exports = cors(async (req, res) => {
  try {
    const [initialUrl] = req.url.replace('/', '').match(/http.+$/) || [''];
    const { protocol, hostname, search } = parse(initialUrl);
    if (!protocol || !hostname)
      throw createError(500, `Invalid url: ${initialUrl}`);

    // Function to request data from Rest API.
    const request = query =>
      axios({
        method: 'get',
        url: `${protocol}//${hostname}/${query || ''}`,
        headers: {
          'user-agent': req.headers['user-agent'],
          host: hostname,
        },
        responseType: 'json',
      });

    // Function to return data with title populated.
    const getModifiedData = data => {
      const { title } = data;

      if (typeof title === 'string') {
        return { ...data, title: { rendered: title, text: title } };
      }

      if (typeof title === 'object') {
        const { rendered, text } = title;

        if (rendered)
          return {
            ...data,
            title: { rendered, text: rendered.replace(/<\/?[^>]+(>|$)/g, '') },
          };

        if (text) return { ...data, title: { rendered: text, text } };
      }

      return data;
    };

    // Get data from original call.
    const { data, headers } = await request(search);

    if (typeof data === 'string')
      throw createError(500, `Invalid url: query is missing`);

    // Preserve x-wp headers
    Object.entries(headers)
      .filter(([key]) => /x-wp/i.test(key))
      .forEach(([key, value]) => res.setHeader(key, value));

    // Check if data is a list of entities with titles.
    if (Array.isArray(data))
      return send(
        res,
        200,
        data.map(entity => (entity.title ? getModifiedData(entity) : entity)),
      );

    // Check if data is an entity with title.
    if (data.title) return send(res, 200, getModifiedData(data));

    return send(res, 200, data);
  } catch (error) {
    const status =
      error.statusCode || (error.response && error.response.status) || 500;

    return send(res, status, {
      error: {
        message: error.message,
        status,
      },
    });
  }
});
